### ISSUE TITLE
External book: Sorry, that book had no pages

### ISSUE THREAD

**Issue by bobnl** [Simbul/baker#264]
Hello all!

I have a problem with using a external book. After I confirmed that I want to download the external book, this error shows up: "Sorry, that books had no pages" (in the IOS-simulator)

This is what I have done:
- Include a link in the default/local book: < a h r e f="book://subdomain.tld/folder/folder2">img< / a >
- Put the files on the server: subdomain.tld/folder/folder2.hpub/(book content, HTML, images, book.json...)
- Changed the URL in the book.json file on the server:     "url": "book://subdomain.tld/folder/folder2/",

Does anyone know why this doesn't work? What am I doing wrong?

Many Thanks!

**Comment by folletto** [comment]
Usually, the problem that triggers that error message is that you zipped the _folder_ instead of its content. :)

The difference is that to zip the folder you right click the folder and zip it.

To zip the content you have to select all the files inside the folder and zip them.

Check if this solves the issue. :)

**Comment by bobnl** [comment]
Thanks for your answer! The files where in a folder on the server, not zipped. Now I have zipped the files and renamed the zip it to book.hpub. Unfortunately this gives another error: "There was a problem downloading the book".

Do you know what the problem is?

Thanks!

**Comment by folletto** [comment]
Yup: Hpub must be zipped. :)

Try this:
1. Take the link from the HTML page that should download the file, for example: "book://subdomain.tld/folder/book"
2. Replace "book://" with "http://": "http://subdomain.tld/folder/book"
3. Add ".hpub": "http://subdomain.tld/folder/book.hpub"
4. Copy that URL in a browser and see if the browser can download the file.

:)

**Comment by bobnl** [comment]
Yes, the browser can download the hpub.

**Comment by folletto** [comment]
Could you copy/paste the **exact** tag code you have in the HTML, replacing the domain so we can't take it, just to see the syntax?

**Comment by bobnl** [comment]
< a href="book://dev.domain.com/comics/comic1" >

**Comment by folletto** [comment]
Ok, that's fine. Uhm...

I'd say to try looking at the Xcode console log and the Apache log as well to try checking if there are mismatches.

**Comment by bobnl** [comment]
Yes, good idea. Is it possible that the hpub is not accessible because it's a sub-domain?

2011-12-30 19:06:23.061 Baker[1517:f803] ? Download file http://dev.domain.com/comics/comic1.hpub
[Switching to process 1517 thread 0xf803]
wait_fences: failed to receive reply: 10004003
2011-12-30 19:06:31.028 Baker[1517:f803] HTTP Request to http://dev.domain.com/comics/comic1.hpub
2011-12-30 19:06:31.631 Baker[1517:f803] Connection received response
2011-12-30 19:06:31.635 Baker[1517:f803] Response status code: 404
2011-12-30 19:06:31.636 Baker[1517:f803] Connection failed! Error - server returned status code 404
2011-12-30 19:06:31.761 Baker[1517:f803] ? Error while downloading new book data

**Comment by folletto** [comment]
Check Apache as well, because it's receiving a response, but it's a 404, it means that Apache is responding... from Apache log you should be able to get another piece of information.

**Comment by Xm4s** [comment]
That's quite strange, let's see ...

Could you please put your .hpub on your local machine and see if the link works in your local browser, Safari Mobile and then Baker?
Assuming that you want to mantain the folder hierachy:
• the link to download the package via browser should be "http://address-to-local-machine/comics/comic1.hpub"
• the correct link for baker should be "book://address-to-local-machine/comics/comic1"

Ah, another things, maybe a little useless, but ...
Assuming you are using a Mac check your book package info (cmd+i) and see if the file name is "comic1.hpub" and not "comic1.hpub.zip" ... I've fallen for this more than once, so I think it's worth taking a look :P

**Comment by bobnl** [comment]
xx.xx.xx.xxx - - [30/Dec/2011:19:06:31 +0100] "GET /comics/comic1.hpub/ HTTP/1.1" 404 480 (simulator)
xx.xx.xx.xxx - - [30/Dec/2011:19:25:27 +0100] "GET /comics/comic1.hpub HTTP/1.1" 304 177 (browser)

The "/" at the end of the URL seems to be the problem. Any idea why this happens?

**Comment by bobnl** [comment]
@Xm4s I try, thanks!

**Comment by bobnl** [comment]
It's working!! Now I use a "normal" domainname instead of a sub-domain, that seems to be the solution.

**Comment by folletto** [comment]
Wow, that's interesting. We need to do some test ourselves now. :o

**Comment by Simbul** [comment]
This is most likely going to depend on the webserver configuration.
I did some testing. What I found out is:
- subdomain names do not seem to influence download
- a trailing slash, however, results in a 404 both in the browser and (obviously) in Baker

A trailing slash in a URL is usually meant to indicate a folder, so it does not make much sense after the path to an HPub file. 

The behaviour associated with the trailing slash is dependent on the webserver, but I wouldn't be surprised if the majority of them would fail to find the content for an HPub url with a trailing slash.

Anyway @bobnl, unless you have any other questions or comments, I think we can consider the issue closed, right? :)
